teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,372 @@
+3. CEL would do the replacement with 1 and 2 with built-in variable substitution
+
+
+### Validation
+
+The new fields will be gated by `cel-in-whenexpression`, which will default to false while the feature is in alpha.
+
+The validation webhook will validate
+1. The feature flag is enabled
+2. If the flag is enabled, either the current when input+operator+values or cel is used, users cannot use both at the same time for one WhenExpression, but they can use different syntax for different WhenExpressions.
+3. If the CEL expression is valid.
+
+### Example
+
+This is an E2E example:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  generateName: guarded-pr-
+spec:
+  pipelineSpec:
+    params:
+      - name: path
+        type: string
+        description: The path of the file to be created
+      - name: branches
+        type: array
+        description: The list of branch names
+    workspaces:
+      - name: source
+        description: |
+          This workspace is shared among all the pipeline tasks to read/write common resources
+    tasks:
+      - name: create-file
+        when:
+          - cel: "'$(params.path)' == 'README.md'"
+        workspaces:
+          - name: source
+            workspace: source
+        taskSpec:
+          workspaces:
+            - name: source
+              description: The workspace to create the readme file in
+          steps:
+            - name: write-new-stuff
+              image: ubuntu
+              script: 'touch $(workspaces.source.path)/README.md'
+      - name: check-file
+        params:
+          - name: path
+            value: "$(params.path)"
+        workspaces:
+          - name: source
+            workspace: source
+        runAfter:
+          - create-file
+        taskSpec:
+          params:
+            - name: path
+          workspaces:
+            - name: source
+              description: The workspace to check for the file
+          results:
+            - name: exists
+              description: indicates whether the file exists or is missing
+          steps:
+            - name: check-file
+              image: alpine
+              script: |
+                if test -f $(workspaces.source.path)/$(params.path); then
+                  printf yes | tee $(results.exists.path)
+                else
+                  printf no | tee $(results.exists.path)
+                fi
+      - name: echo-file-exists
+        when:
+          - cel: "'$(tasks.check-file.results.exists)' in ['yes']"
+        taskSpec:
+          steps:
+            - name: echo
+              image: ubuntu
+              script: 'echo file exists'
+    finally:
+      - name: finally-task-should-be-executed
+        when:
+          - cel: "'$(tasks.echo-file-exists.status)' == 'Succeeded'"
+          - cel: "'$(tasks.status)' ==  'Succeeded'"
+          - cel: "'$(tasks.check-file.results.exists)' == 'yes'"
+          - cel: "'$(params.path)' == 'README.md'"
```

Comment:
@Yongxuanzhang this example is using params and result, but the requirement says that they won't be supported, could you please clarify?

Reply:
I have requested to clarify the requirement - https://github.com/tektoncd/community/pull/1074/files#r1357605404

Reply:
Thanks @pritidesai! I have updated the requirement to clarify. We can do variable replacements, but users cannot set a CEL expression for params.